### PR TITLE
Added whitelist of file types in file upload

### PR DESCRIPTION
**Pull request by khavinshankar**
## Proposed Changes

- Added whitelist of file types in file upload (currently allowing images, videos, audios, pdf, and docs)

### Associated Issue

- wrt Beagal security audit report

## Merge Checklist

- [x] Tests added/fixed
- [ ] Update docs in `/docs`
- [x] Linting Complete
- [x] Any other necessary step

_*Only PR's with test cases included and passing lint and test pipelines will be reviewed*_

@coronasafe/care-backend-maintainers @coronasafe/care-backend-admins

### CODE REVIEW

**Review by Ashesh3 — APPROVED** [review]
LGTM
The whitelist seems really narrow. We can add more as the requirements roll in.
Also, should this be moved to the [`FileUploadCreateSerializer`](https://github.com/coronasafe/care/blob/fileupload-whitelist/care/facility/api/serializers/file_upload.py#L126) serializer? @sainak

**Review by rithviknishad — COMMENTED** [review]
This would prevent users from uploading DICOM images right?

**Review by rithviknishad — APPROVED** [review]

**Review by sainak — APPROVED** [review]

File: care/facility/api/serializers/file_upload.py [hunk 1]
```diff
@@ -120,19 +122,28 @@ class Meta:
             "signed_url",
             "internal_name",
             "original_name",
+            "mime_type",
         )
-        write_only_fields = ("associating_id",)
+        write_only_fields = ("associating_id", "mime_type")
```
Comment (vigneshhari) [inline care/facility/api/serializers/file_upload.py:127]: We can remove this.

File: care/facility/api/serializers/file_upload.py [hunk 2]
```diff
@@ -120,19 +122,28 @@ class Meta:
...
 
     def create(self, validated_data):
         user = self.context["request"].user
+        mime_type = validated_data["mime_type"]
```
Comment (vigneshhari) [inline care/facility/api/serializers/file_upload.py:131]: use pop to remove and fetch at the same time.

File: care/facility/api/serializers/file_upload.py [hunk 3]
```diff
@@ -120,19 +122,28 @@ class Meta:
...
+
+        if mime_type not in settings.ALLOWED_MIME_TYPES:
+            raise ValidationError("Invalid File Type")
```
Comment (vigneshhari) [inline care/facility/api/serializers/file_upload.py:134]: Validation Error should be a json maybe something like ```{"detail" : "Invalid FIle Type"}```

Comment (vigneshhari) [inline care/facility/api/serializers/file_upload.py:134]: Use the Validation error from the DRF package.

File: config/settings/base.py [hunk 4]
```diff
@@ -509,6 +509,45 @@
     default=f"https://{FILE_UPLOAD_BUCKET}.s3.amazonaws.com",
 )
 
+ALLOWED_MIME_TYPES = env.json(
+    "ALLOWED_MIME_TYPES",
+    default=[
```
Comment (vigneshhari) [inline config/settings/base.py:514]: Either switch this to a `list` or a `tuple` method from `env` or make the default value a json object.

File: care/facility/models/file_upload.py [hunk 5]
```diff
@@ -83,14 +83,17 @@ def save(self, *args, **kwargs):
...
 
-    def signed_url(self, duration=60 * 60):
+    def signed_url(self, duration=60 * 60, mime_type=None):
         s3Client = boto3.client("s3", **cs_provider.get_client_config())
+        params = {
+            "Bucket": settings.FILE_UPLOAD_BUCKET,
+            "Key": f"{self.FileType(self.file_type).name}/{self.internal_name}",
+        }
+        if mime_type:
+            params["ContentType"] = mime_type
```
Comment (vigneshhari) [inline care/facility/models/file_upload.py:93]: Have you checked with a live S3 instance and tried to send an invalid payload? If so what was the response?

**Review by vigneshhari — CHANGES_REQUESTED** [review]
Changes Required

File: care/facility/models/file_upload.py (hunk 5 above)
Comment (khavinshankar) [reply care/facility/models/file_upload.py:93]: When we set a content type and upload a different file (with a different mime type), the file is getting uploaded to the bucket, but it's not rendered in the iframe when fetched.

Comment (khavinshankar) [reply care/facility/models/file_upload.py:93]: @vigneshhari 

I tried using generate_presigned_post too, still, we can bypass the ContentType value set while creating the pre-signed URL, any suggestions to make this work? As a last option, we could proxy the file content through the backend!

File: care/facility/api/serializers/file_upload.py [hunk 6]
```diff
@@ -120,19 +122,27 @@ class Meta:
...
+            "mime_type",
         )
         write_only_fields = ("associating_id",)
 
     def create(self, validated_data):
         user = self.context["request"].user
+        mime_type = validated_data.pop("mime_type")
+
+        if mime_type not in settings.ALLOWED_MIME_TYPES:
+            raise ValidationError({"detail": "Invalid FIle Type"})
```
Comment (vigneshhari) [inline care/facility/api/serializers/file_upload.py:134]:
```suggestion
            raise ValidationError({"detail": "Invalid File Type"})
```

**Review by vigneshhari — APPROVED** [review]

**Pull request merged**